src/core/lib/event_engine/windows/iocp.cc

```diff
@@ -74,43 +73,36 @@ Poller::WorkResult IOCP::Work(EventEngine::Duration timeout,
   DWORD bytes = 0;
   ULONG_PTR completion_key;
   LPOVERLAPPED overlapped;
-  if (GRPC_TRACE_FLAG_ENABLED(grpc_event_engine_trace)) {
-    gpr_log(GPR_DEBUG, "IOCP::%p doing work", this);
-  }
+  GRPC_EVENT_ENGINE_POLLER_TRACE("IOCP::%p doing work", this);
   BOOL success = GetQueuedCompletionStatus(
       iocp_handle_, &bytes, &completion_key, &overlapped,
       static_cast<DWORD>(Milliseconds(timeout)));
   if (success == 0 && overlapped == nullptr) {
-    if (GRPC_TRACE_FLAG_ENABLED(grpc_event_engine_trace)) {
-      gpr_log(GPR_DEBUG, "IOCP::%p deadline exceeded", this);
-    }
+    GRPC_EVENT_ENGINE_POLLER_TRACE("IOCP::%p deadline exceeded", this);
     return Poller::WorkResult::kDeadlineExceeded;
   }
   GPR_ASSERT(completion_key && overlapped);
   if (overlapped == &kick_overlap_) {
-    if (GRPC_TRACE_FLAG_ENABLED(grpc_event_engine_trace)) {
-      gpr_log(GPR_DEBUG, "IOCP::%p kicked", this);
-    }
+    GRPC_EVENT_ENGINE_POLLER_TRACE("IOCP::%p kicked", this);
     outstanding_kicks_.fetch_sub(1);
     if (completion_key == (ULONG_PTR)&kick_token_) {
       return Poller::WorkResult::kKicked;
     }
     grpc_core::Crash(
         absl::StrFormat("Unknown custom completion key: %lu", completion_key));
   }
-  if (GRPC_TRACE_FLAG_ENABLED(grpc_event_engine_trace)) {
-    gpr_log(GPR_DEBUG, "IOCP::%p got event on OVERLAPPED::%p", this,
-            overlapped);
+  GRPC_EVENT_ENGINE_POLLER_TRACE("IOCP::%p got event on OVERLAPPED::%p", this,
+                                 overlapped);
+  auto* socket = reinterpret_cast<WinSocket*>(completion_key);
+  if (socket->IsShutdown()) {
```

Comment:
Let's check if the access to socket needs to be synchronized.
Consider cases of the time-of-check-to-time-of-use type

Reply:
Recap of the discussion: the concern here is if `socket->IsShutdown()` return false, but the socket is then shut down in some other thread while either A) the WinSocket callback is being run, or B) the subsequent IOCP::Work call deletes the WinSocket. 

This is valid, and I think points to the need for a ref counted WinSocket type. I'll need to think through it. Unfortunately at the moment, aside from crashes and UB, I don't think we have a good way to test for this, and it's probably a fairly rare occurrence.